src/lobpcg/algorithm.rs

```diff
@@ -0,0 +1,658 @@
+use crate::{cholesky::*, eigh::*, norm::*, triangular::*};
+///! Locally Optimal Block Preconditioned Conjugated
+///!
+///! This module implements the Locally Optimal Block Preconditioned Conjugated (LOBPCG) algorithm,
+///which can be used as a solver for large symmetric eigenproblems.
+use crate::{LinalgError, Result};
+use ndarray::prelude::*;
+use ndarray::{concatenate, Data, ScalarOperand};
+use num_traits::{Float, NumCast};
+
+use std::iter::Sum;
+
+/// Find largest or smallest eigenvalues
+#[derive(Debug, Clone)]
+pub enum Order {
+    Largest,
+    Smallest,
+}
+
+/// The result of the eigensolver
+///
+/// In the best case the eigensolver has converged with a result better than the given threshold,
+/// then a `LobpcgResult::Ok` gives the eigenvalues, eigenvectors and norms. If an error ocurred
+/// during the process, it is returned in `LobpcgResult::Err`, but the best result is still returned,
+/// as it could be usable. If there is no result at all, then `LobpcgResult::NoResult` is returned.
+/// This happens if the algorithm fails in an early stage, for example if the matrix `A` is not SPD
+#[derive(Debug)]
+pub enum LobpcgResult<A> {
+    Ok(Array1<A>, Array2<A>, Vec<A>),
+    Err(Array1<A>, Array2<A>, Vec<A>, LinalgError),
+    NoResult(LinalgError),
+}
+
+/// Solve the generalized eigenvalue problem with pencil (A, B)
+fn generalized_eig<S: Data<Elem = A>, A: NdFloat>(
+    a: ArrayBase<S, Ix2>,
+    b: ArrayBase<S, Ix2>,
+) -> Result<(Array1<A>, Array2<A>)> {
+    let (vals_b, vecs_b) = b.eigh()?;
+    let vals_b_recip = vals_b.mapv(|x| (x.max(A::from(1e-10f32).unwrap())).sqrt().recip());
+    let vecs_b_tilde = &vecs_b * &vals_b_recip;
+    let a_tilde = vecs_b_tilde.t().dot(&a.dot(&vecs_b_tilde));
+    let (vals_a, vecs_a) = a_tilde.eigh()?;
+    let vecs = vecs_b_tilde.dot(&vecs_a);
+
+    Ok((vals_a, vecs))
+}
+
+/// Solve full eigenvalue problem, sort by `order` and truncate to `size`
+fn sorted_eig<S: Data<Elem = A>, A: NdFloat>(
+    a: ArrayBase<S, Ix2>,
+    b: Option<ArrayBase<S, Ix2>>,
+    size: usize,
+    order: &Order,
+) -> Result<(Array1<A>, Array2<A>)> {
+    let n = a.len_of(Axis(0));
+
+    let res = match b {
+        Some(b) => generalized_eig(a, b)?,
+        _ => a.eigh()?.sort_eig(false),
+    };
+
+    // sort and ensure that signs are deterministic
+    let (vals, vecs) = res.sort_eig(false);
+    let s = vecs.row(0).mapv(|x| x.signum());
+    let vecs = vecs * &s;
```

Comment:
Use `vecs.mul_assign(&s)` or `vecs *= &s`. Also, you're multiplying `vecs` by the sign of its first row (this means the first row will always be positive). This is intended right?

Reply:
yes, so that the sign is deterministic in comparisons! You  have one degree of freedom there for eig/SVD

Reply:
removing the borrow let's LLVM reuse the underlying buffer again